services/device-connection/src/main/java/org/eclipse/hono/deviceconnection/infinispan/RemoteCacheBasedDeviceConnectionService.java

```diff
@@ -86,28 +84,36 @@ protected final void doStart(final Promise<Void> startPromise) {
     @Override
     protected void doStart(final Future<Void> startFuture) {
 
-        final Promise<Void> result = Promise.promise();
-        result.future().setHandler(startFuture);
-
         if (cacheManager == null) {
-            result.fail(new IllegalStateException("cache manager is not set"));
+            startFuture.fail(new IllegalStateException("cache manager is not set"));
         } else {
-            context.executeBlocking(r -> {
-                cacheManager.start();
-                cache = cacheManager.getCache("device-connection");
-                cache.start();
-                r.complete(cacheManager);
-            }, attempt -> {
-                if (attempt.succeeded()) {
-                    LOG.info("successfully connected to remote cache");
-                } else {
-                    LOG.info("failed to connect to remote cache", attempt.cause());
-                }
-            });
-            result.complete();
+            connectToGrid();
+            startFuture.complete();
         }
     }
 
+    private void connectToGrid() {
+
+        context.executeBlocking(r -> {
+            if (!cacheManager.isStarted()) {
+                log.debug("trying to start cache manager");
+                cacheManager.start();
+                log.info("started cache manager, now connecting to remote cache");
+            }
+            log.debug("trying to connect to remote cache");
+            cache = cacheManager.getCache("device-connection");
+            cache.start();
+            r.complete(cacheManager);
+        }, attempt -> {
+            if (attempt.succeeded()) {
+                log.info("successfully connected to remote cache");
+            } else {
+                log.debug("failed to connect to remote cache: {}", attempt.cause().getMessage());
```

Comment:
A failed connection to a crucial components should IMHO: a) log to at least "INFO", b) log the full stack trace, not only the message.

Reply:
In general, I agree. To me it is still unclear how the client establishes the connection to the grid and in particular, how it behaves if the connection is lost some time after it has been established initially. The current code explicitly takes care of initial connection but currently does not help with re-establishing a lost connection. Do you have any insight into this? Will the client try to re-connect on itself? Will it endlessly try to re-establish the connection?

Reply:
> Will the client try to re-connect on itself? Will it endlessly try to re-establish the connection?

While not having any knowledge about the code, I learned this from discussions. So yes, once this is started, it should continue to re-connect.